### PR TITLE
[apps] Added CSV stats logging and file output

### LINKED ISSUE #241: Statistics callback for file transfert
Hi,

srt-file-transmit is very useful and efficient for file transfert, but it is not possible to easily get real time information concerning the transfert (bytes send, data lost...).

2 ways could be interesting
 - feed back in the output of the command,
 - callback for a custom field...

I thing the second option is the best for letting user creating a custom script for getting the feedback...

Regards

C. Mourad Jaber

### PR DESCRIPTION
**Pull request by maxlovic**
- Added CSV output format for SRT performance statistics.
- Added file output for SRT performance statistics.

Fixes #241 

**TODO:**
- [x] Added timeout rescheduling on disconnect in timeout mode 1.
- [x] Add `--version` option to print SRT library version info.
- [x] Update `stransmit.md` documentation and include CSV output format and stats file.
- [x] **srt-live-transmit**: `-b:<bandwidth>` option is not parsed (to be deleted)
- [ ] Move `PrintSrtStats()` to `logsupport.cpp` or `apputil.hpp`?
- [ ] Do we need `loginternal` option? The name should be different, because internal log switches on `live-transmit`'s logging function.
- [ ] Show [git revision](https://bravenewmethod.com/2017/07/14/git-revision-as-compiler-definition-in-build-with-cmake/) in the version info of `srt-live-transmit`?

### CODE REVIEW

File: apps/srt-file-transmit.cpp [hunk 1]
```diff
@@ -42,14 +42,27 @@ written by
...
-bool Download(UriParser& srt, UriParser& file);
+bool Upload(UriParser& srt, UriParser& file, std::ostream &out_stats);
+bool Download(UriParser& srt, UriParser& file, std::ostream &out_stats);
 
 const srt_logging::LogFA SRT_LOGFA_APP = 10;
 
 static size_t g_buffer_size = 1456;
 static bool g_skip_flushing = false;
 
+map<string, string> g_options;
```
Comment (ethouris) [inline apps/srt-file-transmit.cpp]: Why did you try to copy the old `Option` functionality from stransmit? It's going to be abandoned and in some further development branches are already deleted. Please use the new one.

File: apps/srt-live-transmit.cpp [hunk 2]
```diff
@@ -182,6 +182,7 @@ int main( int argc, char** argv )
         cerr << "\t-r:<report-frequency=0> - bandwidth report frequency\n";
         cerr << "\t-s:<stats-report-freq=0> - frequency of status report\n";
         cerr << "\t-pf:<format> - printformat (json or default)\n";
+        cerr << "\t-statsreport:<filename> - stats report file name (cout for output to cout, or a filename)\n";
```
Comment (ethouris) [inline apps/srt-live-transmit.cpp]: This option is about defining a file for stats output; wouldn't `-sf` and `-statsfile` sound better?

File: apps/srt-live-transmit.cpp [hunk 3]
```diff
@@ -572,7 +590,7 @@ int main( int argc, char** argv )
                     std::shared_ptr<bytevector> pdata = dataqueue.front();
                     if (!tar.get() || !tar->IsOpen()) {
                         lostBytes += (*pdata).size();
-                    } else if (!tar->Write(*pdata)) {
+                    } else if (!tar->Write(pdata->data(), pdata->size(), out_stats)) {
```
Comment (ethouris) [inline apps/srt-live-transmit.cpp]: What's the reason of having data/size passed separately? It doesn't make a difference, except that you have to pass two of them.

Also, the `Read` and `Write` functions are predicted to implement a single entry point for reading and writing; adding more virtual functions demolishes the design. I personally think that there's nothing wrong in having this `out_stats` parameter written into a global variable, like many other parameters.

File: apps/apputil.hpp [hunk 4]
```diff
@@ -177,6 +196,13 @@ struct OutString
 };
 
 
+struct OutBool
+{
+    typedef bool type;
+    static type process(const options_t::mapped_type& i) { return CheckTrue(i); }
+};
+
+
```
Comment (maxlovic) [inline apps/apputil.hpp]: @ethouris Added `OutBool` option type. To review.

File: apps/apputil.hpp [hunk 5]
```diff
@@ -221,9 +247,16 @@ inline options_t ProcessOptions(char* const* argv, int argc, std::vector<OptionS
...
         {
-            current_key = a+1;
+            string key(a + 1);  // omit '-'
+            size_t pos = key.find_first_of(":=");
+            if (pos == string::npos)
+                pos = key.find(' ');
+            string value = pos == string::npos ? "" : key.substr(pos + 1);
+            key = key.substr(0, pos);
+
+            current_key = key;
```
Comment (maxlovic) [inline apps/apputil.hpp]: @ethouris Added handling of `-option:value` syntax. To review.

File: apps/srt-live-transmit.cpp (hunk 3 above)
Comment (maxlovic) [reply apps/srt-live-transmit.cpp]: This was done mainly because of the file transmit uses the same preallocated buffer in a loop.
```
vector<char> buf(cfg.chunk_size);
```
While `live-transmit` allocates a new buffer every time, which is not very good.
```
std::shared_ptr<bytevector> pdata(
                            new bytevector(cfg.chunk_size));
```
That is why passing a container is not a good approach, because the size of the container may be larger, than the size of the valid data contained.

> Also, the Read and Write functions are predicted to implement a single entry point for reading and writing; adding more virtual functions demolishes the design.

I would then delete the function with the container passed.

> I personally think that there's nothing wrong in having this out_stats parameter written into a global variable

Global variables are hard to follow, when passed from one module to another via several nested functions.

File: apps/srt-file-transmit.cpp (hunk 1 above)
Comment (maxlovic) [reply apps/srt-file-transmit.cpp]: Moved to the new one.

File: apps/srt-live-transmit.cpp (hunk 2 above)
Comment (maxlovic) [reply apps/srt-live-transmit.cpp]: Renamed to `statsout:[stdout|filename]`

File: docs/stransmit.md [hunk 6]
```diff
@@ -164,15 +164,21 @@ example, **-t:60**. Alternatively you can also separate them by a space,
...
+- **-timeout, -t, -to** - Sets the timeout for any activity from any medium (in seconds). Default is 0 for infinite (that is, turn this mechanism off). The mechanism is such that the SIGALRM is set up to be called after the given time and it's reset after every reading succeeded. When the alarm expires due to no reading activity in defined time, it will break the application. **Notes:**
     - The alarm is set up after the reading loop has started, **not when the application has started**. That is, a caller will still wait the standard timeout to connect, and a listener may wait infinitely until some peer connects; only after the connection is established is the alarm counting started. 
     - **The timeout mechanism doesn't work on Windows at all.** It behaves as if the timeout was set to **-1** and it's not modifiable.
+- **-timeout-mode, -tm** - timeout mode used. Default is 0 - timeout will happen after the specified time. Mode 1 cancels the timeout if the connection was established.
```
Comment (ethouris) [inline docs/stransmit.md]: Unclear and misleading. Why not introduce an option for "connection timeout" instead? Moreover, it should be better explained how it applies to listener and the caller because for caller you have also `conntimeo` option.

File: docs/stransmit.md [hunk 7]
```diff
@@ -164,15 +164,21 @@ example, **-t:60**. Alternatively you can also separate them by a space,
...
+- **-bandwidth, -bitrate, -b** [**To be added**] - slow down sending to not send more data than given bandwidth (in bits per second). Note that it has nothing to do with SRT-controlled bandwidth; rather a possibility to have a "live-alike" reading speed from an infinite-speed input source (such as file). Better to think of it as for "testing purposes" because the real live transmission requires its speed synchronized with TS timestamps and *srt-live-transmit* has no ability to interpret them.
 - **-verbose, -v** - display additional information on the standard output. Note that it's not allowed to be combined with output specified as **file://con**
-- **-bandwidth-report, -bitrate-report, -report, -r** - Throw periodic bandwidth reports on the standard output. The value is the number of sending events between reports.
-- **-crash, -k** - Force the program to "crash" (terminate abnormally with core dump) when a problem occurred (even such as connection got broken). This is useful only for developers.
+- **-bandwidth-report, -bitrate-report, -report, -bwreport, -r** - Show periodic bandwidth reports on the standard output. The value is the number of sending events between reports.
```
Comment (ethouris) [inline docs/stransmit.md]: AFAIR this option has been removed, or at least was planned to be removed. It doesn't make any sense because this slowdown doesn't come together with transport stream timestamps. It was for testing purposes only, but even in this form it's limited.

File: docs/stransmit.md [hunk 8]
```diff
@@ -164,15 +164,21 @@ example, **-t:60**. Alternatively you can also separate them by a space,
...
 - **-loglevel** - lowest logging level for SRT, one of: *fatal, error, warning, note, debug* (default: *error*)
 - **-logfa** - selected FAs in SRT to be logged (default: all is enabled, that is, you can filter out log messages comong from only wanted FAs using this option)
-- **-ttl, -max-loss-delay** - use `SRTO_LOSSMAXTTL` option for all SRT medium (applies delayed sending of loss report, if UDP packet reordering is detected, in the number of packets to be received after the lost one before reporting a loss; the real "reorder tolerance" value is controlled internally, starts from 0, and grows only upon detected packet reordering; this option only defines its maximum value; the sensible value is 10)
+- **-ttl, -max-loss-delay** [**To be added**] - use `SRTO_LOSSMAXTTL` option for all SRT medium (applies delayed sending of loss report, if UDP packet reordering is detected, in the number of packets to be received after the lost one before reporting a loss; the real "reorder tolerance" value is controlled internally, starts from 0, and grows only upon detected packet reordering; this option only defines its maximum value; the sensible value is 10)
```
Comment (ethouris) [inline docs/stransmit.md]: Don't add this, rather remove this description. It was probably short living and then removed; this can be well enough used with `lossmaxttl` option.

File: apps/apputil.hpp [hunk 9]
```diff
@@ -221,9 +247,16 @@ inline options_t ProcessOptions(char* const* argv, int argc, std::vector<OptionS
     {
         const char* a = *p;
         // cout << "*D ARG: '" << a << "'\n";
-        if ( moreoptions && a[0] == '-' )
+        if (moreoptions && a[0] == '-')
         {
-            current_key = a+1;
+            string key(a + 1);  // omit '-'
+            size_t pos = key.find_first_of(":=");
```
Comment (ethouris) [inline apps/apputil.hpp]: Using `=` as an alternative value specification is rather not a good idea, especially if the argument might contain spaces. Note that we handle this `:` only for backward compatibility, the desired way to specify values for options is to use the next argument.

File: apps/apputil.hpp (hunk 5 above)
Comment (ethouris) [reply apps/apputil.hpp]: Rest of the things looks good.

File: apps/srt-file-transmit.cpp [hunk 10]
```diff
@@ -59,110 +51,172 @@ void OnINT_ForceExit(int)
...
+        PrintOptionHelp(o_logfa, "<fas=general,...>", "log functional area [all,general,bstats,control,data,tsbpd,rexmit]");
+        PrintOptionHelp(o_logfile, "<filename="">", "write logs to file");
+        PrintOptionHelp(o_quiet, "", "quiet mode (default off)");
+        PrintOptionHelp(o_verbose, "", "verbose mode (default off)");
+        cerr << "\n";
+        cerr << "\t-h,-help - show this help\n";
+        cerr << "\t-version - print SRT library version\n";
+        cerr << "\n";
+        cerr << "\t<input-uri>  - [srt://<ip>:<port>?<query>,udp://<ip>:<port>?<query>,file:///<path>,file://con]\n";
+        cerr << "\t<output-uri> - [srt://<ip>:<port>?<query>,udp://<ip>:<port>?<query>,file:///<path>,file://con]\n";
```
Comment (ethouris) [inline apps/srt-file-transmit.cpp]: Very long lines. I think you better explain these arguments commonly as URI, giving syntax requirements and schemes supported.

File: apps/srt-file-transmit.cpp [hunk 11]
```diff
@@ -606,6 +668,110 @@ bool Download(UriParser& srt_source_uri, UriParser& fileuri)
...
+        {
+            UDT::setlogstream(logfile_stream);
+        }
+    }
+
+    //
+    // SRT stats output
+    //
+    std::ofstream logfile_stats; // leave unused if not set
+    if (cfg.stats_out != "" && cfg.stats_out != "stdout")
```
Comment (ethouris) [inline apps/srt-file-transmit.cpp]: "stdout" potentially may be a filename. It's better to use some symbol or some well known standard, at least used by this application. For consistency, I'd simply parse this value through URIParser, with "path" expected on lack of scheme, so stdout will be then specified as `file://con`.

File: apps/srt-file-transmit.cpp [hunk 12]
```diff
@@ -606,6 +668,110 @@ bool Download(UriParser& srt_source_uri, UriParser& fileuri)
...
+    {
+        logfile_stats.open(cfg.stats_out.c_str());
+        if (!logfile_stats)
+        {
+            cerr << "ERROR: Can't open '" << cfg.stats_out << "' for writing stats. Fallback to stdout.\n";
+            logfile_stats.close();
```
Comment (ethouris) [inline apps/srt-file-transmit.cpp]: You don't have to close a file that failed to open.

File: apps/srt-file-transmit.cpp [hunk 13]
```diff
@@ -606,6 +668,110 @@ bool Download(UriParser& srt_source_uri, UriParser& fileuri)
...
+    //
+    // SRT stats output
+    //
+    std::ofstream logfile_stats; // leave unused if not set
+    if (cfg.stats_out != "" && cfg.stats_out != "stdout")
+    {
+        logfile_stats.open(cfg.stats_out.c_str());
+        if (!logfile_stats)
+        {
+            cerr << "ERROR: Can't open '" << cfg.stats_out << "' for writing stats. Fallback to stdout.\n";
```
Comment (ethouris) [inline apps/srt-file-transmit.cpp]: Not sure if fallback to stdout is a good idea. Maybe better to simply exit the application with error message.

File: apps/apputil.hpp (hunk 4 above)
Comment (ethouris) [reply apps/apputil.hpp]: This looks ok.

File: docs/stransmit.md (hunk 6 above)
Comment (maxlovic) [reply docs/stransmit.md]: Might be a good alternative. Although this `-tm` option is already in master after PR #570. 
But the idea with `timeout-mode` was also to add timeout mode 2, that will start after the connection is established. This will be useful for automated testing.
I suggest to make a separate issue for that, if we want to have another name. This PR is to clean up the command-line parsing and add CSV stats.

File: apps/srt-file-transmit.cpp (hunk 10 above)
Comment (maxlovic) [reply apps/srt-file-transmit.cpp]: Changed to:
```
cerr << "\t<input-uri>  - [scheme://HOST:PORT/PATH?PARAM1=VALUE&PARAM2=VALUE]\n";
cerr << "\t<output-uri> - [scheme://HOST:PORT/PATH?PARAM1=VALUE&PARAM2=VALUE]\n";
cerr << "\tSupported schemes: srt, udp, file://con for stdout\n";
```

File: apps/apputil.hpp (hunk 9 above)
Comment (maxlovic) [reply apps/apputil.hpp]: Deleted `=`

File: docs/stransmit.md (hunk 8 above)
Comment (maxlovic) [reply docs/stransmit.md]: Deleted.

File: docs/stransmit.md (hunk 7 above)
Comment (maxlovic) [reply docs/stransmit.md]: Good. Deleted from the docs.

File: apps/srt-file-transmit.cpp (hunk 13 above)
Comment (maxlovic) [reply apps/srt-file-transmit.cpp]: Added return statement.

File: apps/srt-file-transmit.cpp (hunk 11 above)
Comment (maxlovic) [reply apps/srt-file-transmit.cpp]: Removed `stdout` value from options. Default to stdout anyway. Check if `file://con` is specified as well.

File: apps/srt-file-transmit.cpp (hunk 10 above)
Comment (ethouris) [reply apps/srt-file-transmit.cpp]: Better, but I see no point in repeating the scheme twice. Simply announce that they both use the URI syntax, which would be explained just once.

File: docs/stransmit.md (hunk 6 above)
Comment (ethouris) [reply docs/stransmit.md]: Ok, so let's leave it for later then.

File: apps/srt-live-transmit.cpp (hunk 3 above)
Comment (ethouris) [reply apps/srt-live-transmit.cpp]: If you leave just one version of Read and Write functions, this should be good enough.

File: apps/srt-file-transmit.cpp (hunk 10 above)
Comment (maxlovic) [reply apps/srt-file-transmit.cpp]: Accepted.
